File: flutter-im-ui/lib/src/models/conversation_row_data.dart
import '../components/flare_avatar.dart';

/// Neutral, presentational data for one inbox row — the spec's `ConversationRow`
/// data type. The host maps a core conversation-list-view item into this; all
/// product-specific formatting (emoji/sticker inlining, i18n, group sender
/// prefix) is resolved upstream into the plain [preview] string.
class ConversationRowData {
  const ConversationRowData({
    required this.id,
    required this.title,
    this.avatarUrl,
    this.preview = '',
    this.timestampLabel = '',
    this.unreadCount = 0,
    this.pinned = false,
    this.muted = false,
    this.mentioned = false,
    this.draftPreview,
    this.presence,
  });

  /// Stable conversation id — also seeds the avatar fallback colour.
  final String id;

  /// Display title.
  final String title;

  /// Optional avatar image URL.
  final String? avatarUrl;

  /// Pre-formatted last-message preview (plain text).
  final String preview;

  /// Pre-formatted timestamp label (e.g. "14:32", "昨天").
  final String timestampLabel;

  /// Unread message count; `0` hides the badge.
  final int unreadCount;

  final bool pinned;
  final bool muted;

  /// Whether the current user is @-mentioned in the latest unread.
  final bool mentioned;

  /// When non-null/non-empty, shown with a "草稿" accent instead of [preview].
  final String? draftPreview;

  /// Optional presence for the avatar dot.
  final FlarePresence? presence;

  bool get hasUnread => unreadCount > 0;
  bool get hasDraft => draftPreview != null && draftPreview!.isNotEmpty;
}

File: ios-im-ui/Sources/FlareIMUI/Models/ConversationRowData.swift
import Foundation

/// Tone for a small inline row tag (group / bot / official / mention markers).
public enum FlareTagTone: Sendable {
    case info
    case warning
    case neutral
}

/// A small inline label rendered next to a conversation title (e.g. "Group",
/// "Bot", "Official", "@"). Product decides the text/tone; the kit renders it.
public struct ConversationRowTag: Identifiable, Sendable {
    public let id: String
    public let text: String
    public let tone: FlareTagTone

    public init(text: String, tone: FlareTagTone = .neutral) {
        self.id = text
        self.text = text
        self.tone = tone
    }
}

/// Neutral, presentational data for one inbox row — the spec's `ConversationRow`
/// data type. The host maps a core conversation-list-view item into this; all
/// product-specific formatting is resolved upstream into `preview`.
public struct ConversationRowData: Identifiable, Sendable {
    public let id: String
    public let title: String
    public let avatarURL: String?
    public let preview: String
    public let timestampLabel: String
    public let unreadCount: Int
    public let pinned: Bool
    public let muted: Bool
    public let mentioned: Bool
    public let draftPreview: String?
    public let presence: FlarePresence?
    /// Inline title tags (group / role / mention). Empty by default.
    public let tags: [ConversationRowTag]

    public init(
        id: String,
        title: String,
        avatarURL: String? = nil,
        preview: String = "",
        timestampLabel: String = "",
        unreadCount: Int = 0,
        pinned: Bool = false,
        muted: Bool = false,
        mentioned: Bool = false,
        draftPreview: String? = nil,
        presence: FlarePresence? = nil,
        tags: [ConversationRowTag] = []
    ) {
        self.id = id
        self.title = title
        self.avatarURL = avatarURL
        self.preview = preview
        self.timestampLabel = timestampLabel
        self.unreadCount = unreadCount
        self.pinned = pinned
        self.muted = muted
        self.mentioned = mentioned
        self.draftPreview = draftPreview
        self.presence = presence
        self.tags = tags
    }

    public var hasUnread: Bool { unreadCount > 0 }
    public var hasDraft: Bool { !(draftPreview ?? "").isEmpty }
}
